File: example/java/src/test/java/com/example/ReflektTest.java
package com.example;

import static org.hamcrest.Matchers.empty;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.not;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.assertTrue;
import static org.reflekt.ReflektBuilder.reflekt;

import java.util.Set;

import org.junit.Test;
import org.reflekt.Reflekt;

public class ReflektTest {

    @Test
    public void createInstance() {
        // given
        Reflekt reflekt = reflekt();

        // when
        Set<String> allTypes = reflekt.getAllTypes();

        // then
        assertNotNull(allTypes);
        assertThat(allTypes, is(not(empty())));
        assertTrue(allTypes.contains(ReflektTest.class.getCanonicalName()));
    }
}

File: src/test/kotlin/se/jensim/reflekt/RefleKtTest.kt
package se.jensim.reflekt

import com.nhaarman.mockito_kotlin.doReturn
import com.nhaarman.mockito_kotlin.mock
import org.hamcrest.Matchers.equalTo
import org.junit.Assert.assertThat
import java.lang.annotation.Inherited
import kotlin.annotation.AnnotationTarget.CLASS
import kotlin.test.Test

internal class RefleKtTest {

    private val target = RefleKt ()

    @Test
    fun `classes annotated with`() {
        val annotatedClasses = target.getClassesAnnotatedWith(ThreeAnnotation::class.java)

        val expected: Set<Class<*>> = setOf(TestSuperInterface::class)
                .map { it.java }.toSet()
        assertThat(annotatedClasses, equalTo(expected))
    }

    @Test
    fun `inherited annotation`() {
        val annotatedClasses = target.getClassesAnnotatedWith(TwoAnnotationWithInheritance::class.java)

        val expected: Set<Class<*>> = setOf(
                TestSuperDuperClass::class,
                TestSuperClass::class,
                TestLeafClass::class)
                .map { it.java }.toSet()
        assertThat(annotatedClasses, equalTo(expected))
    }

    @Test
    fun `transitive subclasses`() {
        val superClasses = target.getSubClasses(TestSuperInterface::class.java)

        val expected: Set<Class<*>> = setOf(
                TestInterface::class,
                TestSuperDuperClass::class,
                TestSuperClass::class,
                TestLeafClass::class)
                .map { it.java }.toSet()
        assertThat(superClasses, equalTo(expected))
    }
}

@Target(CLASS)
internal annotation class OneAnnotation

@OneAnnotation
internal class TestLeafClass : TestSuperClass()

@OneAnnotation
internal abstract class TestSuperClass : TestSuperDuperClass()

@TwoAnnotationWithInheritance
internal abstract class TestSuperDuperClass : TestInterface

internal interface TestInterface : TestSuperInterface

@ThreeAnnotation
internal interface TestSuperInterface

@Inherited
@Target(CLASS)
internal annotation class TwoAnnotationWithInheritance

@Target(CLASS)
internal annotation class ThreeAnnotation

